SERVER-98272 Create remediation for if the timestamp embedded in the bucket's ID does not match the control.min timestamp

seristof opened the pull request:
Tested on the following cases:
1. HELP ticket case repro
2. validate_timeseries_id_timestamp.js
3. Checking that it doesn't overwrite extended range timestamps (where it is a known issue and it is ok for the timestamp embedded in the bucket's ID does not match the control.min timestamp; it will not cause an warning/error on validation versions with SERVER-97441, so a description was added about how this script should only be run on versions that have validation with SERVER-97441)

Additional improvements:

- Added some minor changes to documentation for recently-added bucket version mismatch to clarify the difference in errors vs. warnings in validate for v8.1+ and prior to v8.1 respectively.
- Changed the bucket version mismatch and mixed schema remediation to use` {background: true}` during validation.

4hp-4int commented on embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js:
```diff
@@ -0,0 +1,220 @@
+}
+
+print(
+    '\nScript successfully fixed have a mismatched embedded bucket id timestamp and control.min timestamp!');
```
Remove `have` here (I think).  This log line also implies (semantically) that it only fixed a single bucket.

4hp-4int reviewed (commented):
Overall, the script worked for me, although I had difficulty validating exactly what changed. My feedback for the script at a high level would be to make the logged output more meaningful, and slightly less noisy if possible. Perhaps only logging when we detect an actual mismatch.

dhly-etc commented on embedded-timestamp-mismatch/README.md:
```diff
@@ -0,0 +1,199 @@
+- Detect if the bucket has mismatched embedded bucket ID timestamp and control.min timestamp
+- Re-insert the measurements of the problematic bucket transactionally.
+  - Unpack the measurements.
+  - Insert the measurements back into the collection. These will go into a new buckets.
+  - Delete the problematic bucket from the collection.
```
[nit] Is it clearer if it's worded as
```suggestion
  - Repack the measurements into new buckets.
  - Delete the original, problematic bucket from the collection.
```
?

dhly-etc commented on embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js:
```diff
@@ -0,0 +1,220 @@
+// 2) Re-insert the measurements of the timestamp-mismatched bucket
+// transactionally.
+//    a) Unpack the measurements
+//    b) Insert the measurements back into the collection. These will go into
+//    new buckets.
+//    c) Delete the mismatched embedded bucket id timestamp and
```
[nit] If we update the wording in the README, we should also update it here.

seristof commented:
> Overall, the script worked for me, although I had difficulty validating exactly what changed. My feedback for the script at a high level would be to make the logged output more meaningful, and slightly less noisy if possible. Perhaps only logging when we detect an actual mismatch.

Thanks for the feedback! Deleted some redundant log lines and changed the validate to be with `{background: true}`.

henrikedin commented on embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js:
```diff
@@ -0,0 +1,213 @@
+    const oidTimestamp = bucket._id.getTimestamp();
+    const controlMinTimestamp = bucket.control.min.t
+
+    if (oidTimestamp != controlMinTimestamp) {
```
This should take extended range dates into account. We shouldn't reinsert all data with extended range.

seristof replied:
Done, and added testing with using extended range docs from timeseries_filter_extended_range.js and ensuring that they don't get overwritten.

henrikedin commented on embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js:
```diff
@@ -0,0 +1,213 @@
+// Prior to v8.1, buckets that have a mismatched embedded bucket id timestamp
+// and control.min timestamp will lead to a warning during validation.
+//
+if (validateRes.errors.length != 0 || validateRes.warnings.length != 0) {
```
Can we compare the actual error/warning output? I don't want unrelated warnings to report that there still exist buckets with mismatched _id

seristof replied:
Added checking for the correct log id for bucket version mismatch as well.

DamianWasilewicz commented on embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js:
```diff
@@ -0,0 +1,223 @@
+
+    // If this collection has extended-range measurements, we cannot assert that
+    // the minTimestamp matches the embedded timestamp.
+    print(controlMinTimestamp == oidTimestamp)
```
If we want to keep this log in for each measurement, should we maybe add more detail to it? More-so than just True/False.

seristof replied:
Sorry, this was a debugging print I forgot to remove 😅 . Thanks for pointing it out!

DamianWasilewicz reviewed (approved):
The script itself looks good to me! Nice work Stephanie
